File: src/viewer.py
import sys
sys.path.append("../src")

from PyQt4 import QtCore
from PyQt4 import QtGui
from PyQt4 import QtOpenGL
from OpenGL import GLU
from OpenGL.GL import *

from obj import load_obj
from quaternion import rotate_pt

class ViewerWidget(QtOpenGL.QGLWidget):
    def __init__(self, ofile, parent = None):
        super(QtGui.QWidget, self).__init__(parent)
        self.setFixedWidth(600)

        self.faces, self.vertices = load_obj(ofile)

        self.cam_pos = 5
 
        self.mouse_start = (0,0)
        self.mouse_end = (0,0)
        self.mouse_old = (0,0)

        self.pitch = 0
        self.old_pitch = 0

        self.yaw = 0
        self.old_yaw = 0
   

    def initializeGL(self):
        glClearColor(1,1,1,1)
        glEnable(GL_DEPTH_TEST)
        
        glCullFace(GL_BACK)
        glEnable(GL_CULL_FACE)
        
		
    def resizeGL(self, width, height):
        if height == 0: height = 1
        glMatrixMode(GL_PROJECTION)
        glLoadIdentity();
        GLU.gluPerspective(45, (1.0 * width) / height, 0.1, 20)
        glViewport(0, 0, width, height)

        glMatrixMode(GL_MODELVIEW)
        glLoadIdentity()  


    def paintGL(self):
        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
        glLoadIdentity()
       
        cam = rotate_pt([self.cam_pos, 0, 0], [0,1,0], self.pitch)
        cam = rotate_pt(cam, [0,0,1], self.yaw)

        up = rotate_pt([0,1,0],[0,1,0],self.pitch)
        up = rotate_pt(up, [0,0,1], self.yaw)
        GLU.gluLookAt(cam[0],cam[1],cam[2], 0,0,0, up[0],up[1],up[2])


        for f in self.faces:
            glColor(1, 0, 0)
            p = self.begin_param(f)

            glBegin(p)
            map(glVertex, f)
            glEnd()

            glPolygonMode(GL_FRONT, GL_LINE)
            glDepthFunc(GL_LEQUAL)
            glColor(0,0,0)
            
            glBegin(p)
            map(glVertex, f)
            glEnd()
            glPolygonMode(GL_FRONT,GL_FILL)
            glDepthFunc(GL_LESS)


    def begin_param(self,face):
        l = len(face)
        if l == 3:
            return GL_TRIANGLES
        elif l == 4:
            return GL_QUADS
        
        return GL_POLYGON

    def save_angles(self):
        self.change_angles()
        self.old_pitch = self.pitch
        self.old_yaw = self.yaw

        return self.pitch, self.yaw
    
    def change_angles(self):
        if self.mouse_old == self.mouse_end:
            return

        startx,endx = self.mouse_start[0], self.mouse_end[0]
        starty,endy = self.mouse_start[1], self.mouse_end[1]

        dx = endx - startx
        dy = endy - starty

        self.pitch = (self.old_pitch - dx) % 360
        self.yaw = (self.old_yaw - dy) % 360

        self.mouse_old = self.mouse_end
        return self.pitch, self.yaw


    def mousePressEvent(self, e):
        p = e.pos()
        self.mouse_start = p.x(), p.y()


    def mouseReleaseEvent(self, e):
        p = e.pos()
        self.mouse_end = p.x(), p.y()
        
        self.save_angles()

        self.updateGL()

    
    def mouseMoveEvent(self, e):
        p = e.pos()
        self.mouse_end = p.x(), p.y()
        self.change_angles()
        self.updateGL()


    def wheelEvent(self, e):
        self.cam_pos += e.delta() / 100.0
        self.cam_pos = min(self.cam_pos, 20)
        self.cam_pos = max(self.cam_pos, 0.1)
        self.updateGL()


class MainWindow(QtGui.QMainWindow):
    def __init__(self, ofile):
        QtGui.QMainWindow.__init__(self)

        self.resize(800, 600)
        self.setWindowTitle('Mesh Simplification Project (atbechan)')
        
        self.viewer = ViewerWidget(ofile, self)
        self.setCentralWidget(self.viewer)

        self.initActions()
        
        
    def initActions(self):
        self.exitAction = QtGui.QAction('Quit', self)
        self.exitAction.setStatusTip('Exit application')
        self.connect(self.exitAction, QtCore.SIGNAL('triggered()'), self.close)





if __name__ == "__main__":
    if len(sys.argv) != 2:
        print """usage: viewer FILE"""

        sys.exit(1)

    obj_file = sys.argv[1]

    app = QtGui.QApplication(sys.argv)
    win = MainWindow(obj_file)
    win.show()
    sys.exit(app.exec_())

File: src/viewer.cpp
#include <vector>
#include <GL/glut.h>
#include <GL/glu.h>
#include "viewer.h"
#include "obj.h"
#include "quaternion.h"

#include <iostream>
using namespace std;

Viewer::Viewer(){
    cam_pos = 4;
    mouse_start_x = 0;
    mouse_start_y = 0;

    mouse_end_x = 0;
    mouse_end_y = 0;

    mouse_old_x = 0;
    mouse_old_y = 0;

    pitch = 0;
    old_pitch = 0;

    yaw = 0;
    old_yaw = 0;
}

Viewer::~Viewer(){}

void Viewer::init(char *fname, bool lines){
    obj = new ObjFile(fname);
    drawLines = lines;

    computeDepth();
    computeCenter();

    glClearColor(1,1,1,1);
    glEnable(GL_DEPTH_TEST);

    glCullFace(GL_BACK);
    glEnable(GL_CULL_FACE);
}

void Viewer::resize(int width, int height){
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(45, (1.0 * width) / height, 0.1, 20);
    glViewport(0, 0, width, height);

    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();

}

void Viewer::display(){

    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    glLoadIdentity();

    vector<float> cam(3,0);
    cam[0] = cam_pos;

    vector<float> y_axis(3,0);
    y_axis[1] = 1;

    vector<float> z_axis(3,0);
    z_axis[2] = 1;

    cam = rotatePoint(cam, y_axis, pitch);
    cam = rotatePoint(cam, z_axis, yaw);

    vector<float> up(3,0);
    up[1] = 1;
    up = rotatePoint(up, y_axis, pitch);
    up = rotatePoint(up, z_axis, yaw);
    


    gluLookAt(cam[0], cam[1], cam[2], 0, 0, 0, up[0], up[1], up[2]);



    glTranslatef(-center[0], -center[1], -center[2]);
    glColor3f(1,0,0);
    for(unsigned int i = 0; i < obj->numFaces(); i++){
        vector<vector<float> > verts = obj->getFaceVertices(i);
        GLenum begin = getBeginParam(verts.size());
        
        //draw the shape with a filled color
        glColor3f(1,0,0);
        glBegin(begin);
        {
            vector<vector<float> >::iterator it;
            for(it = verts.begin(); it != verts.end(); it++){
                vector<float> v = *it;
                glVertex3f(v[0], v[1], v[2]);
            }
            glEnd();
        }

        if(drawLines){
            //draw the shape with an outline
            glPolygonMode(GL_FRONT, GL_LINE);
            glDepthFunc(GL_LEQUAL);
            glColor3f(0,0,0);
            glBegin(begin);
            {    
                vector<vector<float> >::iterator it;
                for(it = verts.begin(); it != verts.end(); it++){
                    vector<float> v = *it;
                    glVertex3f(v[0], v[1], v[2]);
                }
            }
            glEnd();
            glPolygonMode(GL_FRONT,GL_FILL);
            glDepthFunc(GL_LESS);
        }

    }
    
    glutSwapBuffers();
}

//mouse click
void Viewer::mouse(int button, int state, int x, int y){
    //for mouse click
    if(button == 3){
        cam_pos -= 0.1f;
        glutPostRedisplay();
        return;
    }
    else if(button == 4){
        cam_pos += 0.1f;
        glutPostRedisplay();
        return;
    }



    //for left click
    int pressed = 0;
    if(state == pressed){
        mouse_start_x = x;
        mouse_start_y = y; 
    }
    else{
        mouse_end_x = x;
        mouse_end_y = y;

        saveAngles();
    }

    glutPostRedisplay();
}

//mouse motion
void Viewer::mouse(int x, int y){
    mouse_end_x = x;
    mouse_end_y = y;
    changeAngles();
    glutPostRedisplay();
}

void Viewer::changeAngles(){
    if(mouse_old_x == mouse_end_x && mouse_old_y == mouse_end_y){
        return;
    }

    int dx = mouse_end_x - mouse_start_x;
    int dy = mouse_end_y - mouse_start_y;

    pitch = (old_pitch - dx) % 360;
    yaw = (old_yaw - dy) % 360;

    if(pitch < 0)
        pitch += 360;

    if(yaw < 0)
        yaw += 360;

    mouse_old_x = mouse_end_x;
    mouse_old_y = mouse_end_y;
}

void Viewer::saveAngles(){
    changeAngles();
    old_pitch = pitch;
    old_yaw = yaw;
}


GLenum Viewer::getBeginParam(unsigned int i){
    if(i == 3){
        return GL_TRIANGLES;
    }
    else if(i == 4){
        return GL_QUADS;
    }
    return GL_POLYGON;
}


void Viewer::computeDepth(void){
    float m = 0;
    for(unsigned int i = 0; i < obj->numVertices(); i++){
        vector<float> v = obj->getVertex(i);

        if(v[0] > m)
            m = v[0];
        if(v[1] > m)
            m = v[1];
        if(v[2] > m)
            m = v[2];
    }


    depth = m;
    cam_pos = depth + 4;
}


void Viewer::computeCenter(void){
    center.push_back(0);
    center.push_back(0);
    center.push_back(0);

    unsigned int size = obj->numVertices();
    for(unsigned int i = 0; i < size; i++){
        vector<float> v = obj->getVertex(i);
        center[0] += v[0];
        center[1] += v[1];
        center[1] += v[2];
    }


    center[0] /= (float) size;
    center[1] /= (float) size;
    center[2] /= (float) size;

}
